src/html/data-sections.js

```diff
@@ -0,0 +1,177 @@
+ */
+function findPlaceholders(section, handlefn) {
+  visit(section, (node) => {
+    if (node.value && pattern.test(node.value)) {
+      handlefn(node, 'value');
+    }
+    if (node.alt && pattern.test(node.alt)) {
+      handlefn(node, 'alt');
+    }
+    if (node.url && pattern.test(node.url)) {
+      handlefn(node, 'url');
+    }
+    if (node.title && pattern.test(node.title)) {
+      handlefn(node, 'title');
+    }
+  });
+}
+
+/**
+ * Determines if an MDAST node contains placeholders like `{{foo}}`
+ * @param {MDAST} section
+ */
+function hasPlaceholders(section) {
+  try {
+    findPlaceholders(section, () => {
+      throw new Error('Placeholder detected');
+    });
+    return false;
+  } catch {
+    return true;
+  }
+}
+
+/**
+ * @param {MDAST} section
+ */
+function fillPlaceholders(section) {
+  if (!section.meta || (!section.meta.embedData && !Array.isArray(section.meta.embedData))) {
+    return;
+  }
+  const data = section.meta.embedData;
+  // required to make deepclone below work
+  removePosition(section);
+
+  const children = data.reduce((p, value) => {
+    const workingcopy = deepclone(section);
+
+    findPlaceholders(workingcopy, (node, prop) => {
+      if (typeof node[prop] === 'string') {
+        node[prop] = node[prop].replace(pattern, (_, expr) => dotprop.get(value, expr));
+      }
+    });
+    return [...p, ...workingcopy.children];
+  }, []);
+
+  section.children = children;
+  delete section.meta.embedData;
+}
+
+function normalizeLists(section) {
+  function cleanupLists([first, second]) {
+    // two consecutive, identical lists
+    if (first.type === 'list' && second.type === 'list'
+        && first.ordered === second.ordered
+        && first.start === second.start
+        && first.spread === second.spread) {
+      // move the children of the first to the second list
+      second.children = [...first.children, ...second.children];
+      first.children = [];
+      // mark the first list to be emptied
+      return first;
+    }
+    return null;
+  }
+
+  // get the sequence of lists to be removed
+  const emptiedLists = pipe(
+    trySlidingWindow(section.children, 2),
+    map(cleanupLists),
+    reject(is(null)),
+    list,
+  );
+
+  // perform the cleanup
+  section.children = pipe(
+    section.children, // take all existing children
+    reject((child) => contains(is(child))(emptiedLists)), // reject if they are in the list above
+    list, // make a nice array because the rest of the world doesn't like iterators yet
+  );
+}
+
+async function fillDataSections({ content: { mdast } }, { downloader }) {
+  async function extractData(section) {
+    return pmap(section, async (node) => {
+      if (node.type === 'dataEmbed') {
+        const task = downloader.getTaskById(`dataEmbed:${node.url}`);
+        const downloadeddata = await task;
+        // TODO: better error handling
+        // TODO: check that the result is an array
+        const json = await downloadeddata.json();
```

Comment:
This doesn't work. `downloadeddata` is not a Fetch API `Response` but a js object: `{ status, body, headers }`

This should work:

```js
const json = JSON.parse(downloadeddata.body);
```